modules/store/src/selector.ts

```diff
@@ -118,393 +118,42 @@ export function defaultMemoize(
-
-export function createSelector<State, S1, S2, S3, S4, S5, S6, S7, Result>(
-  s1: Selector<State, S1>,
-  s2: Selector<State, S2>,
-  s3: Selector<State, S3>,
-  s4: Selector<State, S4>,
-  s5: Selector<State, S5>,
-  s6: Selector<State, S6>,
-  s7: Selector<State, S7>,
-  projector: (s1: S1, s2: S2, s3: S3, s4: S4, s5: S5, s6: S6, s7: S7) => Result
-): MemoizedSelector<State, Result>;
-export function createSelector<
-  State,
-  Props,
-  S1,
-  S2,
-  S3,
-  S4,
-  S5,
-  S6,
-  S7,
-  Result
->(
-  s1: SelectorWithProps<State, Props, S1>,
-  s2: SelectorWithProps<State, Props, S2>,
-  s3: SelectorWithProps<State, Props, S3>,
-  s4: SelectorWithProps<State, Props, S4>,
-  s5: SelectorWithProps<State, Props, S5>,
-  s6: SelectorWithProps<State, Props, S6>,
-  s7: SelectorWithProps<State, Props, S7>,
-  projector: (
-    s1: S1,
-    s2: S2,
-    s3: S3,
-    s4: S4,
-    s5: S5,
-    s6: S6,
-    s7: S7,
-    props: Props
-  ) => Result
-): MemoizedSelectorWithProps<State, Props, Result>;
-export function createSelector<State, S1, S2, S3, S4, S5, S6, S7, Result>(
-  selectors: [
-    Selector<State, S1>,
-    Selector<State, S2>,
-    Selector<State, S3>,
-    Selector<State, S4>,
-    Selector<State, S5>,
-    Selector<State, S6>,
-    Selector<State, S7>
-  ],
-  projector: (s1: S1, s2: S2, s3: S3, s4: S4, s5: S5, s6: S6, s7: S7) => Result
-): MemoizedSelector<State, Result>;
-export function createSelector<
-  State,
-  Props,
-  S1,
-  S2,
-  S3,
-  S4,
-  S5,
-  S6,
-  S7,
-  Result
->(
-  selectors: [
-    SelectorWithProps<State, Props, S1>,
-    SelectorWithProps<State, Props, S2>,
-    SelectorWithProps<State, Props, S3>,
-    SelectorWithProps<State, Props, S4>,
-    SelectorWithProps<State, Props, S5>,
-    SelectorWithProps<State, Props, S6>,
-    SelectorWithProps<State, Props, S7>
-  ],
-  projector: (
-    s1: S1,
-    s2: S2,
-    s3: S3,
-    s4: S4,
-    s5: S5,
-    s6: S6,
-    s7: S7,
-    props: Props
-  ) => Result
-): MemoizedSelectorWithProps<State, Props, Result>;
-
-export function createSelector<State, S1, S2, S3, S4, S5, S6, S7, S8, Result>(
-  s1: Selector<State, S1>,
-  s2: Selector<State, S2>,
-  s3: Selector<State, S3>,
-  s4: Selector<State, S4>,
-  s5: Selector<State, S5>,
-  s6: Selector<State, S6>,
-  s7: Selector<State, S7>,
-  s8: Selector<State, S8>,
-  projector: (
-    s1: S1,
-    s2: S2,
-    s3: S3,
-    s4: S4,
-    s5: S5,
-    s6: S6,
-    s7: S7,
-    s8: S8
-  ) => Result
-): MemoizedSelector<State, Result>;
+type StateOfSelectors<S> = S extends Selector<infer State, any>[]
+  ? State
+  : never;
+
+type ValueOfSelector<S> = S extends Selector<any, infer Value> ? Value : never;
+
+export function createSelector<S extends Selector<any, any>[], Result>(
+  ...args: [
+    ...S,
+    (...args: [...{ [i in keyof S]: ValueOfSelector<S[i]> }]) => Result
+  ]
+): MemoizedSelector<StateOfSelectors<S>, Result>;
 export function createSelector<
-  State,
+  S extends SelectorWithProps<any, any, any>[],
```

Comment:
The old definition of selectors with props specified that all input selectors had those props, but that doesn't quite make sense to me. Couldn't they be any arbitrary selectors, and the `Props` are merely inferred from the projector?